modules/core/src/main/java/org/apache/ignite/internal/util/distributed/DistributedProcess.java

```diff
@@ -0,0 +1,388 @@
+import static org.apache.ignite.internal.managers.communication.GridIoPolicy.SYSTEM_POOL;
+
+/**
+ * Distributed process is a cluster-wide process that accumulates single nodes results to finish itself.
+ * <p>
+ * The process consists of the following phases:
+ * <ol>
+ *  <li>Initial request starts process. (Sent via discovery)</li>
+ *  <li>Each server node process an initial request and send the single node result to the coordinator. (Sent via
+ *  communication)</li>
+ *  <li>The coordinator accumulate all single nodes results and finish process. (Sent via discovery)</li>
+ * </ol>
+ * <p>
+ * Several processes of one type can be started at the same time.
+ *
+ * @param <I> Request type.
+ * @param <R> Result type.
+ */
+public class DistributedProcess<I extends Serializable, R extends Serializable> {
+    /** Kernal context. */
+    private final DistributedProcessType type;
+
+    /** Active processes. */
+    private final ConcurrentHashMap<UUID, Process> processes = new ConcurrentHashMap<>(1);
+
+    /** Synchronization mutex for coordinator initializing and the remaining collection operations. */
+    private final Object mux = new Object();
+
+    /** Kernal context. */
+    private final GridKernalContext ctx;
+
+    /** Logger. */
+    private final IgniteLogger log;
+
+    /**
+     * @param ctx Kernal context.
+     * @param type Process type.
+     * @param exec Execute action and returns future with the single node result to send to the coordinator.
+     * @param finish Finish process closure. Called on each node when all single nodes results received.
+     */
+    public DistributedProcess(GridKernalContext ctx, DistributedProcessType type,
+        Function<I, IgniteInternalFuture<R>> exec,
+        CI3<UUID, Map<UUID, R>, Map<UUID, Exception>> finish) {
+        this.ctx = ctx;
+        this.type = type;
+
+        log = ctx.log(getClass());
+
+        ctx.discovery().setCustomEventListener(InitMessage.class, (topVer, snd, msg) -> {
+            if (msg.type() != type)
+                return;
+
+            Process p = processes.computeIfAbsent(msg.processId(), id -> new Process(msg.processId()));
+
+            // May be completed in case of double delivering.
+            if (p.initFut.isDone())
+                return;
+
+            ClusterNode crd = coordinator();
+
+            if (crd == null) {
+                p.initFut.onDone();
+
+                onAllServersLeft();
+
+                return;
+            }
+
+            p.crdId = crd.id();
+
+            if (crd.isLocal())
+                initCoordinator(p, topVer);
+
+            IgniteInternalFuture<R> fut = exec.apply((I)msg.request());
+
+            fut.listen(f -> {
+                if (f.error() != null)
+                    p.resFut.onDone(f.error());
+                else
+                    p.resFut.onDone(f.result());
+
+                ClusterNode crdNode = coordinator();
+
+                if (crdNode != null && !ctx.clientNode())
```

Comment:
Why we just ignore the fact that coordinator unknown?
We should print this in log, at least.
But I think we should throw an exception in that case.

Reply:
I left a check on the client node only. Coordinator can not be null on a server node.